qiskit_experiments/test/mock_iq_backend.py

```diff
@@ -52,32 +61,164 @@ def _default_options(self):
+            state_str = val2str_dict[idx]
+            for _ in range(number_of_occurrences):
+                shot_memory = []
+                for qubit_number, char_qubit in enumerate(state_str):
+                    # the iteration on the str starts from the MSB so we will use a variable to
+                    # make the code more readable.
+                    current_qubit = num_qubits - qubit_number - 1
+                    # The structure of iq_centers is [qubit_number][logic_result][I/Q].
+                    i_center = iq_centers[current_qubit][int(char_qubit)][0]
+                    q_center = iq_centers[current_qubit][int(char_qubit)][1]
+                    point_i = i_center + qubits_iq_rand[shot_num][qubit_number]
+                    point_q = q_center + qubits_iq_rand[shot_num][qubit_number]
+                    
+                    # Adding phase if not 0.0
+                    if not np.allclose(phase, 0.0):
+                        complex_iq = (point_i + 1.0j * point_q) * np.exp(1.0j * phase)
+                        point_i, point_q = np.real(complex_iq), np.imag(complex_iq)
+                    
+                    shot_memory.append([point_i, point_q])
+                # We proceed to the next occurrence - meaning its a new shot.
+                memory.append(shot_memory)
+                shot_num += 1
 
         return memory
 
+    def _generate_data(self, prob: Dict[str, float], num_qubits: int, circuit: QuantumCircuit) -> Dict:
+        """
+
+        Args:
+            prob(dict):
+            num_qubits:
+
+        Returns:
+
+        """
+        # Maybe I need to get as input for generalization
+        shots = self.options.get("shots")
+        meas_level = self.options.get("meas_level")
+        meas_return = self.options.get("meas_return")
+        run_result = {}
+
+        val2str_dict = self._values_to_string_array(num_qubits)
+        if meas_level == MeasLevel.CLASSIFIED:
+            counts = {}
+            results = self._rng.multinomial(shots, prob, size=1)
+            for result, num_occurrences in enumerate(results):
+                result_in_str = val2str_dict["result"]
+                counts[result_in_str] = num_occurrences
+            run_result["counts"] = counts
+        else:
+            phase = self._iq_phase(circuit)
+            memory = self._draw_iq_shots(prob, shots, num_qubits, phase)
+            if meas_return == "avg":
+                memory = np.average(np.array(memory), axis=0).tolist()  # could have a bug here
+
+            run_result["memory"] = memory
+        return run_result
+
     @abstractmethod
-    def _compute_probability(self, circuit: QuantumCircuit) -> float:
+    def _compute_probability(self, circuit: QuantumCircuit) -> Dict[str, float]:
         """Compute the probability used in the binomial distribution creating the IQ shot.
 
         An abstract method that subclasses will implement to create a probability of
```

Comment:
Hi @eggerdj , I wanted to consult with you about removing the abstract function `_compute_probability(self, circuit: QuantumCircuit)`.
I think it will be more comfortable for the user to provide a dictionary that contains the output string (in binary) and its probability to the run function of the backend instead of implementing `_compute_probability(self, circuit: QuantumCircuit)`.
In both cases, the user need to calculate the probabilities, but the current method require to inherit from `MockIQBackend` and implementing `_compute_probability(self, circuit: QuantumCircuit)` which I think is cumbersome.
In the future, we can add a function to calculate the probability for generic circuit as @yaelbh suggested in #532 .

Reply:
Good point. Something to get ride of all of these experiment specific mock backends is a good idea. I'm not convinced that providing a dict is the most general way to go. Perhaps the user can initialize the mock backend with an optional function called `compute_probability` something like:
```python
class MockIQBackend(FakeOpenPulse2Q):
    def __init__(self, ..., compute_probabilities: Optional[Callable] = None):
        ...
        self._compute_probabilities = compute_probabilities or self._aer_wrapper
```
If this function is not given then we obtain the probabilities from the Aer simulator (which cannot reproduce the miscalibration patterns that we have, we would need a pulse simulator for that). Here, both `compute_probabilities` and `_aer_wrapper` should have the signature: 
```python
def _compute_probability(self, circuit: QuantumCircuit) -> Dict[str, float]:
```
or perhaps
```python
def _compute_probability(self, circuit: List[QuantumCircuit]) -> List{Dict[str, float]]:
```
to be more efficient?

Reply:
Yea, I think this solution is the most general as I give the user almost full freedom in making that function (up to the input).
I will do it as you suggested, thanks!

Reply:
I think maybe we need to change the signature to be as following:
```python
def _compute_probability(self, circuit: QuantumCircuit, calc_parameters: Dict[str, Any]) -> Dict[str, float]:
```
This will allow the user to use parameters to calculate the probabilities if necessary, like `ideal_beta` for Drag experiment.
I don't want to enforce this pattern on the user incase he doesn't need those parameters so I thought it could be:
```python
def __init__(
        self,
        iq_cluster_centers: List[Tuple[Tuple[float, float], Tuple[float, float]]] = None,
        iq_cluster_width: List[float] = None,
        rng_seed: int = 0,
        calculation_parameters: Optional[Dict[str, Any]] = None,
        compute_probabilities: Optional[Union[Callable[[Union[QuantumCircuit, List[QuantumCircuit]]],
                                                       Union[Dict[str, float], List[Dict[str, float]]]],
                                              Callable[[Union[QuantumCircuit, List[QuantumCircuit]],
                                                        Optional[Dict, List[Dict]]],
                                                       Union[Dict[str, float], List[Dict[str, float]]]]]] = None,
    ):
```
Do you think this is an efficient solution? The variable declaration is overloaded but I don't know if there is a more elegant way to do it.

Reply:
@ItamarGoldman I think what you're trying with `calc_parameters` is in fact `**kwargs`. And then you probably won't need the monster in `__init__`.